iml-agent/src/daemon_plugins/ostpool.rs

```diff
@@ -118,7 +117,136 @@ impl DaemonPlugin for PoolState {
     fn update_session(
         &self,
     ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
-        // @@ diff
-        future::ok(None).boxed()
+        let state = Arc::clone(&self.state);
+
+        async move {
+            let fslist = list_fs()
+                .await
+                .map_err(|_| Ok::<Vec<String>, ImlAgentError>(vec![]))
+                .unwrap();
+
+            let xs = fslist.into_iter().map(|fs| {
+                let state = Arc::clone(&state);
+
+                async move {
+                    if let Some(hm) = state.lock().await.get_mut(&fs) {
+                        let pools = pools(fs.clone()).await?;
+
+                        let mut rmlist = hm.clone();
+                        let mut changes: Vec<(OstPoolAction, OstPool)> = vec![];
+                        for pool in pools.iter() {
+                            tracing::debug!("Updating {}.{}", fs, pool.name);
+
+                            if let Some(oldosts) = rmlist.remove(&pool.name) {
+                                let new: HashSet<_> = pool.osts.iter().cloned().collect();
+                                let old: HashSet<_> = oldosts.iter().cloned().collect();
+
+                                let addlist: HashSet<_> = new.difference(&old).cloned().collect();
+                                if !addlist.is_empty() {
+                                    hm.insert(
+                                        pool.name.clone(),
+                                        old.union(&addlist).cloned().collect(),
+                                    );
+                                    changes.push((
```

Comment:
This function is really deep, maybe factor out some of the logic to other functions?